### PR TITLE
[maistra-367] [maistra-701] SMCP templates

### PR DESCRIPTION
**Pull request by brian-avery**

### CODE REVIEW

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 1]
```diff
@@ -208,20 +231,116 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
...
+		if _, exists := base[key]; !exists {
+			base[key] = value
+			continue
+		}
+
+		// at this point, key exists in both input and base.
+		// If both are maps, recurse.
+		// If only input is a map, ignore it. We don't want to overrwrite base.
+		// If both are values, again, ignore it since we don't want to overrwrite base.
+		if baseKeyAsMap, baseOK := base[key].(map[string]interface{}); baseOK {
```
Comment (dmitri-d) [inline pkg/controller/servicemesh/controlplane/reconciler.go:237]: Are values in the base config are supposed to be the default values? If so, shouldn't values in the input map take precedence?

File: pkg/controller/servicemesh/controlplane/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
+package controlplane
```
Comment (dmitri-d) [inline pkg/controller/servicemesh/controlplane/controller_test.go:1]: yay, tests!

File: deploy/smcp-templates/product [hunk 3]
```diff
@@ -0,0 +1,37 @@
+apiVersion: maistra.io/v1
```
Comment (dmitri-d) [inline deploy/smcp-templates/product]: nit: should the name of the template be "service-mesh", not "product"?

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 4]
```diff
@@ -49,10 +52,30 @@ var orderedCharts = []string{
 	"istio/charts/kiali",
 }
 
+const (
+	smcpTemplatePath        = "/etc/istio-operator/templates/"
```
Comment (dmitri-d) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: we probably want to keep the templates in ```/usr/share/istio-operator```, as this is an immutable application data, not application config.

Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Just to keep terminology consistent are these `profiles` or `templates`?

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 5]
```diff
@@ -208,20 +231,116 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
...
+		//configmaps mounted in directories with pre-existing content
+		defaultTemplateContent, defaultErr := ioutil.ReadFile(smcpDefaultTemplatePath + name)
+		if defaultErr != nil {
+			return v1.ControlPlaneSpec{}, fmt.Errorf("template cannot be loaded from user or default directory. Error from user: %s. Error from default: %s", err, defaultErr)
+		}
+		templateContent = defaultTemplateContent
+	}
+
+	var template v1.ServiceMeshControlPlane
+	if err = yaml.Unmarshal(templateContent, &template); err != nil {
```
Comment (rcernich) [inline pkg/controller/servicemesh/controlplane/reconciler.go:264]: Should we be using k8s decoder for this?  I believe that's the proper way to deserialize a k8s resource.

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 6]
```diff
@@ -49,10 +52,30 @@ var orderedCharts = []string{
...
+	//render the ServiceMeshControlPlane templates
+	spec, err := r.updateSMCPWithTemplates(r.Instance.Spec)
+	if err != nil {
+		r.Log.Info(fmt.Sprintf("Warning: failed to render ServiceMeshControlPlane templates:%s", err))
+
+		//TODO: Remove these comments we've decided that templates are stable.
+		//updateReconcileStatus(&r.Instance.Status.StatusType, err)
+		//return reconcile.Result{}, err
+	} else {
+		r.Instance.Spec = spec
```
Comment (rcernich) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: We should not be modifying any user specified fields.  We should store the result in the status, as you have below and use that value when rendering the templates.

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 7]
```diff
@@ -208,20 +231,116 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
...
+
+	var template v1.ServiceMeshControlPlane
+	if err = yaml.Unmarshal(templateContent, &template); err != nil {
+		return v1.ControlPlaneSpec{}, fmt.Errorf("failed to parse template %s contents: %s", name, err)
+
+	}
+	return template.Spec, nil
+}
+
+func (r *ControlPlaneReconciler) renderSMCPTemplates(smcp v1.ControlPlaneSpec, visited map[string]struct{}) (v1.ControlPlaneSpec, error) {
```
Comment (rcernich) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: nit, on the naming.  We're not rendering anything here.  Maybe something like, `applyTemplate()`?

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 8]
```diff
@@ -208,20 +231,116 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
...
+	}
+
+	visited[smcp.Template] = struct{}{}
+
+	smcp.Istio = mergeValues(smcp.Istio, template.Istio)
+	smcp.ThreeScale = mergeValues(smcp.ThreeScale, template.ThreeScale)
+	return smcp, nil
+}
+
+func (r *ControlPlaneReconciler) updateSMCPWithTemplates(smcpSpec v1.ControlPlaneSpec) (v1.ControlPlaneSpec, error) {
```
Comment (rcernich) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: nit, naming.  maybe `applyTemplates()`.  conflicts a bit with `applyTemplate()` above, but it seems like this is applying the templates to the smcp resource, while the above is recursive, going through each template referenced by the template.

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 9]
```diff
@@ -208,20 +231,116 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
...
 	r.Log.Info("rendering helm charts")
 	allErrors := []error{}
 	var err error
 	var threeScaleRenderings map[string][]manifest.Manifest
 
-	if globalValues, ok := r.Instance.Spec.Istio["global"].(map[string]interface{}); ok {
-		globalValues["operatorNamespace"] = r.OperatorNamespace
-	} else {
-		return fmt.Errorf("Could not set operatorNamespace value, as .Values.global is not a map[string]interface{}: %v", err)
+	if r.Instance.Spec.Istio == nil {
```
Comment (rcernich) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: I think we should be using `r.Instance.Status.LastAppliedConfiguration` here, as mentioned above.  If we're doing that, I don't think any of these other changes are necessary, other than to operate on the status value instead of the spec.

**Review by rcernich — CHANGES_REQUESTED** [review]

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 9 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: The code between here and 341 is a fix for 701. Currently, if an SMCP is created without an Istio section, the operator will crash every time it goes to process the config because there was no check for nil on the Istio field. Also, if a user is using templates, the following a perfectly valid SMCP:

`apiVersion: maistra.io/v1
kind: ServiceMeshControlPlane
metadata:
  name: basic-install
spec:
  template: maistra`

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 10]
```diff
@@ -49,10 +52,30 @@ var orderedCharts = []string{
...
+	smcpDefaultTemplatePath = "/etc/istio-operator/default-templates/"
+	smcpDefaultTemplate     = "default"
+)
+
 func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
 	var err error
 	var ready bool
 
+	//render the ServiceMeshControlPlane templates
+	spec, err := r.updateSMCPWithTemplates(r.Instance.Spec)
```
Comment (rcernich) [inline pkg/controller/servicemesh/controlplane/reconciler.go]: I think this should be moved into `renderCharts()`.

Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: We should probably only be doing this when we render the charts, which is not on every reconcile, as we may be coming back after pausing for a component to become available.

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 1 above)
Comment (dgn) [reply pkg/controller/servicemesh/controlplane/reconciler.go:237]: I think the naming is a bit misleading, but we're calling this with user-supplied config as `base` and the template config as `input`, so this behavior is correct.

File: deploy/smcp-templates/base [hunk 11]
```diff
@@ -0,0 +1,48 @@
+apiVersion: maistra.io/v1
```
Comment (luksa) [inline deploy/smcp-templates/base:1]: How come this file is called `base` and not `base.yaml`?

Comment (brian-avery) [reply deploy/smcp-templates/base:1]: The filenames are used as the name of the template. If the user decides to create a file named foo.yaml then the template needs to be called foo.yaml. Same with JSON. Since extension doesn't matter in Linux, I removed it to clean things up a bit.

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 12]
```diff
@@ -49,10 +52,30 @@ var orderedCharts = []string{
 	"istio/charts/kiali",
 }
 
+const (
+	smcpTemplatePath        = "/etc/istio-operator/templates/"
+	smcpDefaultTemplatePath = "/etc/istio-operator/default-templates/"
+	smcpDefaultTemplate     = "default"
```
Comment (dgn) [inline pkg/controller/servicemesh/controlplane/reconciler.go:57]: I was wondering if we could use this `default` template to replace our default-overrides that we currently do via `sed`? We'd have to make sure all those defaults end up in the other templates as well, of course.

Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go:57]: That's my hope here. Everything except for a couple of deletes is between base, Maistra, and product. Maistra specific stuff is what was community before. Product stuff is what was product before. Base is everything else. Product and Maistra use base as their template.

Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go:57]: The build process will add either Maistra or product as default.

Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go:57]: I like this idea.  Have you tried using `null` for the values being deleted?

Comment (dgn) [reply pkg/controller/servicemesh/controlplane/reconciler.go:57]: awesome!

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 4 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: I've been using the term templates since this is a template for a user to overlay their custom settings on top of rather than a profile which to me means a complete set of configuration assigned to a specific user.

Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: **profile** --  the personal details, images, user statistics, social media timeline, etc., that an individual creates and associates with a username or online account

**template** -- A template is a file that serves as a starting point for a new document.

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 6 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Fixed

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 10 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Done

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 9 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Fixed

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 8 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Correct. This one is "applying" the template, or updating the SMCP with the associated template values. Whereas the other one traverses all templates before processing them.

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 7 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Done

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 8 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Done

File: deploy/smcp-templates/product (hunk 3 above)
Comment (brian-avery) [reply deploy/smcp-templates/product]: Done

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 5 above)
Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go:264]: This relies on the scheme to create the required resources, which is nil in the controlplane. Leaving as is for now. Created maistra-720 to look into this.

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 4 above)
Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go]: Would something like `pattern` work better?  Template seems a bit overused.

File: pkg/apis/maistra/v1/servicemeshcontrolplane_types.go [hunk 13]
```diff
@@ -44,21 +44,28 @@ type ControlPlaneStatus struct {
...
+	LastAppliedConfiguration ControlPlaneSpec `json:"lastAppliedConfiguration"`
 }
 
 // HelmValuesType is typedef for Helm .Values
 type HelmValuesType map[string]interface{}
 
 // ControlPlaneSpec represents the configuration for installing a control plane
 type ControlPlaneSpec struct {
+	//Template selects the template to use for default values
+	Template string `json:"template,omitempty"`
```
Comment (rcernich) [inline pkg/apis/maistra/v1/servicemeshcontrolplane_types.go:58]: Would it make sense to have this be a list?  For example, a user specifies something like:

```yaml
spec:
  template:
  - full
  - auth
```

Comment (brian-avery) [reply pkg/apis/maistra/v1/servicemeshcontrolplane_types.go:58]: Think we discussed this with Kevin before and decided not to for now. This would also bring up questions like in what order do things overlay each other. Instead we have a hierarchical inheritance.

File: tmp/build/patch-jaeger.sh [hunk 14]
```diff
@@ -1,5 +1,12 @@
 #!/usr/bin/env bash
 
+function jaeger_patch_values() {
+	# update jaeger zipkin port name
+  sed -i -e '/service:$/,/externalPort:/ {
```
Comment (brian-avery) [inline tmp/build/patch-jaeger.sh:5]: @rcernich No idea what this is doing. Doesn't look like it gets reflected in the final values.yaml. Reverted for now.

File: deploy/smcp-templates/servicemesh [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+      tag: 1.0.0
+    sidecarInjectorWebhook:
+      image: sidecar-injector-rhel8
+    gateways:
+      istio-ingressgateway:
+        ior_image: istio-ior-rhel8
+    prometheus:
+      image: prometheus-rhel8
+  threeScale:
+    hub: registry.redhat.io/openshift-istio-tech-preview
```
Comment (philipgough) [inline deploy/smcp-templates/servicemesh]: @brian-avery remove `openshift-istio-tech-preview` and replace with `openshift-service-mesh` here and elsewhere?

File: pkg/apis/maistra/v1/servicemeshcontrolplane_types.go (hunk 13 above)
Comment (rcernich) [reply pkg/apis/maistra/v1/servicemeshcontrolplane_types.go:58]: complication here is that it would require an api change, so better to decide now, as we'll have to live with it for a while.

as for the logic, why not apply them in order, full then auth, given the example above.  nothing complicated about that.

Comment (rcernich) [reply pkg/apis/maistra/v1/servicemeshcontrolplane_types.go:58]: adding some color, this allows for composition, which avoids an expansion of templates (full-auth, minimal-auth, etc.), "prefer composition over inheritance."

File: pkg/controller/servicemesh/controlplane/reconciler.go (hunk 5 above)
Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go:264]: This is nil, https://github.com/Maistra/istio-operator/blob/maistra-1.0/pkg/controller/servicemesh/controlplane/controller.go#L132?  Somehow the scheme is there, because the client uses it to deserialize the control plane object you're working on.  https://github.com/Maistra/istio-operator/blob/maistra-1.0/cmd/manager/main.go#L93

File: pkg/controller/servicemesh/controlplane/reconciler.go [hunk 16]
```diff
@@ -208,34 +216,150 @@ func (r *ControlPlaneReconciler) Reconcile() (reconcile.Result, error) {
...
+	r.Status.LastAppliedConfiguration = spec
+
+	if err := r.validateSMCPSpec(r.Status.LastAppliedConfiguration); err != nil {
+		return err
+	}
 
-	if globalValues, ok := r.Instance.Spec.Istio["global"].(map[string]interface{}); ok {
+	if globalValues, ok := r.Status.LastAppliedConfiguration.Istio["global"].(map[string]interface{}); ok {
 		globalValues["operatorNamespace"] = r.OperatorNamespace
-	} else {
```
Comment (rcernich) [inline pkg/controller/servicemesh/controlplane/reconciler.go:219]: Why did you remove this?  This is required for the cni charts.  If we can't set this, then we can't proceed with the rendering/install.

Comment (rcernich) [reply pkg/controller/servicemesh/controlplane/reconciler.go:219]: ..,or we should proceed as below and create a map for the `global` field.

Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go:219]: @rcernich I moved validation of the global section being present into the SMCP validation (312).

Comment (brian-avery) [reply pkg/controller/servicemesh/controlplane/reconciler.go:219]: r.Status.LastAppliedConfiguration.Istio["global"]

still has the same problem as the previous code if the above layers aren't checked. If .Istio is nil, then attempting to access the global field will crash.

And the global section not being present at all is likely an issue even outside of setting the operatornamespace. I.e. what tag do you use?

File: tmp/build/patch-jaeger.sh [hunk 17]
```diff
@@ -35,9 +22,9 @@ function jaeger_patch_production() {
 function JaegerPatch() {
   echo "Patching Jaeger"
 
-  jaeger_patch_values
```
Comment (rcernich) [inline tmp/build/patch-jaeger.sh:38]: you've removed the invocation, but not the function

Comment (rcernich) [reply tmp/build/patch-jaeger.sh:38]: sorry, didn't see that it was moved below

**Review by rcernich — APPROVED** [review]

File: tmp/build/patch-charts.sh [hunk 18]
```diff
@@ -78,37 +66,17 @@ function patchTemplates() {
...
-    sed -i -e 's/hub:.*$/hub: kiali/' \
-           -e 's/tag:.*$/tag: v'${KIALI_VERSION}'/' ${HELM_DIR}/istio/charts/kiali/values.yaml
-  else
-    sed -i -e 's+hub:.*$+hub: registry\.redhat\.io\/openshift-service-mesh+' \
-           -e 's/tag:.*$/tag: '${KIALI_VERSION}'/' ${HELM_DIR}/istio/charts/kiali/values.yaml
-  fi
-
-  # In Istio 1.2, this viewOnlyMode is there, but in Istio 1.1 we need to add it - it is supported by the latest Kiali
-  sed -i -e 's/grafanaURL:/viewOnlyMode: false\n  grafanaURL:/' ${HELM_DIR}/istio/charts/kiali/values.yaml
+	#CNI is handled separately
```
Comment (dmitri-d) [inline tmp/build/patch-charts.sh:74]: I don't think this is needed anymore -- cni image location is now being specified in the operator deployment ENV.

Comment (brian-avery) [reply tmp/build/patch-charts.sh:74]: Yeah, I'm not sure so I added it back in to be safe. That said, I haven't been able to find anywhere it's actually used.

**Review by dmitri-d — APPROVED** [review]
I think updating location of CNI images can be safely omitted, but it doesn't have affect on the rest of the functionality.

lgtm
